### ISSUE TITLE
[RFC][Ascend NPU][Diffusion] Hybrid W8A8 for Boogu-Image

### ISSUE THREAD

**Issue by QwertyJack** [vllm-project/vllm-omni#6789]
## Motivation

Boogu-Image contains a roughly 19B-parameter execution graph: a diffusion Transformer (DiT), Qwen3-VL-8B-Instruct conditioning model, visual tower, and VAE. The BF16 path peaks at about 36.3-37.0 GiB allocated and 38.2-40.9 GiB reserved on Ascend 910, which is unsuitable for a practical 32 GB 910B4 budget.

This RFC proposes contributing the previously validated Ascend integer quantization work to the Boogu-Image Ascend track in #6665/#6787. It is deliberately separate from the FP8 work in #6677.

## Proposed profile

`Hybrid W8A8 v2` consists of:

| Component | Runtime policy |
|---|---|
| DiT attention Q/K/V/output | W8A16 |
| DiT adaptive modulation | W8A16 |
| DiT FFN gate/up | static W8A8 with one shared input quantization |
| DiT FFN down | dynamic per-token W8A8 after fused SwiGLU |
| Qwen text backbone | official Ascend Qwen W8A8 QuaRot checkpoint |
| Qwen visual tower and VAE | BF16 |

The DiT sidecar profile is `dit-hybrid-w8a8-v2`. Base/Turbo and Edit/Edit-Turbo have different DiT FFN counts (52 and 54 respectively); the Qwen QuaRot checkpoint is shared because the embedded Qwen weights are byte-identical across the four Boogu checkpoints.

## Existing evidence

The local implementation completed 48 independent performance runs (4 models x 4 profiles x 3 repeats) and 128 quality outputs (16 cases x 4 profiles x 2 seeds) on 64 GB Ascend 910:

- Hybrid peak allocated memory: 21.99-22.91 GiB.
- Hybrid peak reserved memory: 23.39-25.77 GiB.
- Hybrid was 1.3-8.9% faster than the DiT-W8A16 + QuaRot capacity profile.
- Base, Turbo, Edit, and Edit-Turbo all completed; 128/128 outputs were finite and usable with no observed functional or visual regression.
- Physical 32 GB 910B4 validation is still pending and is required before claiming 32 GB support.

Detailed local records:

- `rfc/boogu-image-ascend-quantization-overview.md`
- `validation/boogu-hybrid-w8a8-v2-validation-20260722.md`

## Required Omni integration

The upstream prototype replaces modules after loading a complete BF16 pipeline. That is not an acceptable production path for a 32 GB target. Omni needs construction-time quantization and streaming/lazy weight materialization:

1. Pass `quant_config` and stable `prefix` values through `BooguImagePipeline` and `BooguImageTransformer2DModel`.
2. Add a dedicated manifest/config for the Hybrid contract rather than overloading dynamic-activation `DiffusionInt8Config` or generic `--quantization int8`.
3. Implement NPU W8A16 and static/dynamic W8A8 linear methods using existing quantization interfaces and Ascend operators.
4. Fuse DiT FFN expansion, `SiluAndMul`, dynamic per-token activation quantization, and the down projection where the kernel contract permits.
5. After #6649, map the diffusers `to_q/to_k/to_v` and `linear_1/linear_3` checkpoint names to packed QKV/gate-up modules through explicit shard mappings.
6. Adapt Qwen QuaRot as a complete basis: smoothed norms, rotated token embeddings, 216 text W8A8 linears, four rotated visual bridges, and the final inverse block-Hadamard transform. The Qwen `lm_head` is unused and should not be loaded.
7. Route the DiT sidecar and Qwen QuaRot components independently and reject incompatible combinations that modify Qwen twice.

## Deliberate precision boundaries

- Quantizing the Qwen visual tower saved only about 0.256 GiB but reduced image-token cosine to about 0.77; keep it BF16.
- DiT attention activation W8A8 was 15-18% slower in the measured shape; keep attention W8A16.
- Qwen down projections remain at the official FLOAT/BF16 boundary.

## Proposed PR split

- [ ] RFC review and checkpoint/manifest contract.
- [ ] Quant-aware Boogu construction, component routing, and lazy loader.
- [ ] Packed QKV/gate-up checkpoint mapping.
- [ ] DiT Hybrid W8A8 NPU operators and fused FFN.
- [ ] Qwen QuaRot loader/adaptor and output-basis restoration.
- [ ] Four-model real-weight e2e, quality/performance/memory A/B, perf baselines, recipe, supported-models, and expansion e2e.

## Acceptance requirements

- [ ] Real Base, Turbo, Edit, and Edit-Turbo requests succeed through vLLM-Omni; Turbo/Edit-Turbo must use the verified DMD class/revision contract from #6699/#6701.
- [ ] Load-time memory is measured with the target checkpoint; loading a full BF16 graph first is not accepted as 32 GB evidence.
- [ ] Compare against a fixed BF16 reference using image validity, trajectory/numerical probes, pixel metrics, semantic metrics, and visual inspection.
- [ ] Report allocated and reserved memory, latency, and cold-start separately.
- [ ] Complete physical 32 GB Ascend 910B4 validation before marking the target configuration supported.

## Questions for maintainers

1. Should the user-facing method be a new `boogu_hybrid_w8a8`/`ascend_hybrid_w8a8` quantization method, or should the manifest be selected through component-level quantization config?
2. Where should the external Qwen QuaRot checkpoint mapping and revision/hash contract live?
3. Should the first implementation target the existing vLLM-Ascend NPU platform only, with standalone torch-npu support from #6610 handled separately?

**Comment by QwertyJack** [comment]
One loader constraint is decisive for the proposed Hybrid W8A8 implementation:

The current native Boogu pipeline constructs `Qwen3VLForConditionalGeneration` through Transformers `from_pretrained` and moves the inner model to the execution device before the diffusion transformer is loaded (`pipeline_boogu_image.py`). A diffusion `quant_config` therefore cannot quantize the Qwen text component today; it only reaches quantization-aware native modules such as the DiT transformer.

The standalone Ascend prototype applies Qwen QuaRot and DiT sidecars after constructing a complete BF16 pipeline. That is useful evidence, but it is not acceptable as the production 32 GB path because it materializes the full BF16 graph first.

Please choose the first upstream contract explicitly:

1. port Qwen to a native vLLM-Omni encoder with construction-time quantization;
2. add a dedicated Transformers/meta-device loader that materializes QuaRot linears while loading; or
3. publish a preconverted checkpoint whose component configs and serialized scales are understood by the Omni loader.

My recommendation is to land the DiT Hybrid profile and component routing first, then make Qwen QuaRot a separate loader PR. The runtime operators can remain under the existing vLLM-Ascend NPU platform; #6610 standalone torch-npu backend is a separate decision. The method should have an explicit name such as `boogu_hybrid_w8a8`/`ascend_hybrid_w8a8`, not silently reuse generic `int8`.

**Comment by QwertyJack** [comment]
First implementation PR opened as #6791 (draft, stacked on #6649).

It adds only quant-aware Boogu transformer construction: component routing, stable attention/FFN/modulation prefixes, and packed QKV/gate-up mappings. It deliberately does not add a Hybrid W8A8 method, NPU kernels, DiT sidecar loading, or Qwen QuaRot loading.

One mapping detail is now covered explicitly: `instruct_to_qkv` and `img_to_qkv` must precede the shorter `to_qkv` suffix because vLLM packed-module helpers use insertion-ordered suffix matching.

The next proposed PR remains the reviewed manifest/config contract; the dedicated method name and Qwen construction-time loader ownership questions in this RFC are still open.

**Comment by QwertyJack** [comment]
After checking the current quantization implementation in both repositories, I propose the following concrete first contract for maintainer review.

### 1. Reuse the existing component router

vLLM-Omni already has `ComponentQuantizationConfig`, longest-prefix routing, `build_quant_config()`, per-source resolution in `DiffusersPipelineLoader._get_source_quant_config()`, checkpoint adapters, and streaming weight iteration. Hybrid W8A8 should use those interfaces rather than add a second generic routing layer.

The user-facing selection should remain explicit and must not alias generic `int8`, for example:

```json
{
  "transformer": {
    "method": "boogu_hybrid_w8a8",
    "profile": "dit-hybrid-w8a8-v2",
    "manifest": "/path/to/boogu-hybrid-w8a8-v2.json"
  },
  "mllm": {
    "method": "boogu_hybrid_w8a8",
    "profile": "qwen3vl-quarot"
  },
  "vae": null
}
```

The concrete spelling can change, but the `transformer` and `mllm` policies must remain separately addressable and `vae` must remain explicitly BF16.

### 2. Make the manifest self-contained and fail closed

Suggested minimum schema:

```json
{
  "format_version": 1,
  "profile": "boogu-hybrid-w8a8-v2",
  "model_family": "boogu-image-0.1",
  "components": {
    "transformer": {
      "profile": "dit-hybrid-w8a8-v2",
      "source_revision": "...",
      "files": [{"path": "...", "sha256": "..."}]
    },
    "mllm": {
      "profile": "ascend-qwen-w8a8-quarot",
      "model": "Qwen/Qwen3-VL-8B-Instruct",
      "revision": "...",
      "files": [{"path": "...", "sha256": "..."}],
      "load_lm_head": false,
      "visual_policy": "bf16-with-rotated-bridges",
      "output_basis": "inverse-block-hadamard-32"
    },
    "vae": {"dtype": "bfloat16"}
  }
}
```

Loader validation should reject unknown format/profile, missing files, hash mismatch, model-family mismatch, a full SmoothQuant sidecar combined with QuaRot, Qwen visual W8A8, or a second quantization policy targeting the same prefix.

### 3. Use construction-time loading for both heavy components

- **DiT:** construct the packed vLLM linears with the new quant method, then stream serialized qweights/scales through the existing `ComponentSource(prefix="transformer.")` loader. #6791 supplies stable prefixes and packed QKV/gate-up mappings after #6649. Do not construct BF16 `nn.Linear` weights first.
- **Qwen:** add a Boogu/Qwen3-VL construction-time component loader. Instantiate the supported Qwen module structure on `meta`, skip `lm_head`, and materialize each target representation directly while consuming the official QuaRot shards: rotated embedding/norms, 216 text W8A8 linears, four rotated visual bridges, BF16 down projections/visual tower, and the final inverse block-Hadamard restoration. This is not ordinary `Qwen3VLForConditionalGeneration.from_pretrained()` followed by module replacement.

This chooses the preconverted/self-describing checkpoint approach for the first upstream implementation while keeping the runtime module native and auditable. A complete native vLLM encoder can remain a later optimization if reuse justifies it.

### 4. Ownership recommendation

The current vLLM-Omni diffusion `int8_config.py` already owns NPU-specific `NPUInt8LinearMethod`/`NPUInt8OnlineLinearMethod`, while vLLM-Ascend owns reusable AR/general Ascend W8A8/W8A16 schemes. Therefore:

- component config, manifest validation, Boogu/Qwen loader, and the DiT fused-FFN method should first live in **vLLM-Omni**;
- reuse existing vLLM-Ascend schemes where their serialized parameter contract matches, or factor genuinely generic Ascend schemes into vLLM-Ascend in a separate dependency PR;
- do not make the first pure-diffusion implementation depend on the standalone-backend decision in #6610. It must work with the current vLLM-Ascend NPU platform, while keeping the quant method free of AR-only assumptions.

The DiT fused method needs the measured contract that generic dynamic `int8` does not provide: one static shared-input quantization for packed gate/up, `npu_dequant_swiglu_quant(..., quant_mode=1)`, then a down `npu_quant_matmul` with per-token scale. Attention and modulation remain W8A16.

### 5. PR sequence

1. #6649 lands; rebase construction-only #6791 and remove its dependency commit.
2. Manifest/config plus strict validation and synthetic loader tests.
3. DiT serialized Hybrid W8A8 method and fused FFN.
4. Qwen QuaRot construction-time loader and output-basis restoration.
5. Four-model real-weight API/trajectory/quality/perf/memory coverage, followed by physical 32 GB 910B4 sign-off.

@david6666666 @Isotr0py @lishunyang12: please confirm whether this reuse of `ComponentQuantizationConfig` plus a model-specific construction-time Qwen loader is the desired first contract. I will keep #6791 construction-only and draft until #6649 lands; no Hybrid runtime support is being claimed by that PR.

**Comment by hsliuustc0106** [comment]
Measured evidence for the M2 design discussion, from e2e profiling of Boogu-Image Turbo on a single L20X (main @ 13b85c5, vllm 0.29.0, 1024x1024, 4-step DMD):

**The denoise step is already at the BF16 compute roofline, so quantization is the right primary lever.**

- Engine split (stage_0 ~764 ms): 4-step DMD loop ~69% (~186 ms/step), VAE decode + post ~21%, instruction encoding ~5%.
- Joint sequence is ~4,000 tokens (1024 image tokens + ~3,000 system-prompt/template tokens — confirmed by allocation arithmetic on the FFN intermediate), DiT ~18.5B BF16 → ~148 TFLOP per forward.
- Measured GEMM throughput on model-shape GEMMs: ~750 TFLOPS BF16 → compute floor ~197 ms vs measured 186 ms/step. **Within ~6% of the roofline** — launch/dispatch overhead and kernel fusion have little to give at BF16 (two A/B runs with per-block CUDA graphs, dynamic and static, moved stage_0 by <1%; prompt-length A/B moved nothing because the fixed system template dominates the token count).

Implications for the W8A8 contract:

1. **~2× is real and reachable** — on a compute-roofline workload, weight/activation precision maps almost linearly to step time. A measurable target: the 186 ms/step (2.29 s E2E on 910) Turbo row should roughly halve at W8A8 with quality parity.
2. **A second, orthogonal axis nobody has raised: instruction-stream length.** ~3,000 of the ~4,000 tokens are the fixed system prompt/chat template, and DiT FLOPs scale linearly with the sequence. Template compression or encoder-side token reduction cuts step time proportionally (~2.7× if S → 1,500) and composes multiplicatively with W8A8. This is a model/pipeline-side change, independent of quantization format.
3. Not worth pursuing for this model: CUDA graphs / launch-overhead work (measured no-op) and kernel fusion at BF16 (marginal at the roofline).

Full numbers, commands, and the profiling setup can be shared if useful.

**Comment by hsliuustc0106** [comment]
Follow-up to my roofline comment above, with one correction to its anti-recommendations: the **attention backend is worth switching on this stack**. The platform default resolved FLASH_ATTN (FA2, `_vllm_fa2_C::varlen_fwd`); A/B on the same L20X + Turbo server (5-run batches, 1024x1024, 4-step):

| attention backend | stage_0 mean | runs |
|---|---:|---|
| FLASH_ATTN (default) | 762.9 ms | 751.4-777.0 |
| **CUDNN_ATTN** | **672.8 ms** | 666.1-680.6 (**-11.8%**) |

Deterministic (same-seed outputs byte-identical across runs), outputs decode and visually match the FA2 render. TRTLLM_ATTN rejects the card (requires SM100/SM103; the L20X is SM90), FLASH_ATTN_3_HUB has no compatible kernel build for torch 2.13/py3.12, and FLASHINFER isn't a valid selector name on this tree.

Combined with the response-format fix in #7447 (`output_format: "jpeg"`, -134 ms client-side), the full client-observed e2el for Turbo at 1024x1024 drops from **900 ms to 700 ms (-22%)** with flags only — worth documenting in the recipe, and worth a sanity check on whether other diffusion models on SM90 are similarly leaving ~12% attention time on the FA2 default.

**Comment by hsliuustc0106** [comment]
Final review of the contract proposal as the #6665 roadmap author — this is well-constructed, and with the additions below it has my support.

**Agreed without changes:** component-router reuse (§1, no second routing layer), the fail-closed manifest with hashes/revisions (§2), construction-time loading for both heavy components with explicit rejection of post-hoc sidecar application (§3), the vLLM-Omni / vLLM-Ascend ownership split (§4), and the PR sequence (§5).

**Four additions before this is final:**

1. **[Evidence] Report against the BF16 baseline, not only the W8A16+QuaRot profile.** The 1.3-8.9% claim compares two quantized profiles; the roadmap-relevant number is vs BF16. Our kernel-level profile bounds the expectation: GEMMs are 54% of engine time and the W8A8 FFN path covers roughly two-thirds of that, so vs-BF16 step-time gain should land around 15-25% with this mixed profile (attention/modulation stay W8A16) — plus the memory win (36-41 -> 22-26 GiB), which is the primary 32 GB motivation. A per-profile BF16-vs-hybrid table in the same harness closes this.
2. **[Quality gate] Make the 128-output check quantitative.** "No observed functional or visual regression" should become PSNR/SSIM vs the BF16 outputs per case/profile (the repository already ships ffmpeg SSIM/PSNR helpers and the diffusion alignment harness), with fixed seeds and the thresholds stated. The 128 outputs exist; only the metric columns are missing.
3. **[Fallback] Document the option-3 fallback for the Qwen meta-device loader.** The QuaRot meta construction (216 linears + inverse block-Hadamard restoration) is the highest-risk piece of §3. If it proves fragile, a purely preconverted self-describing checkpoint — which the manifest already supports — should be the accepted fallback rather than a design failure.
4. **[Sequencing risk] §5 step 1 depends on #6649, which is still a draft.** No change requested — just flagging that the whole chain (#6649 -> #6791 rebase -> this RFC) is serialized on it, so that rebase is the current critical path for M2.

Also noting the interaction with the roofline measurements posted above: the denoise step is at the BF16 compute roofline with GEMMs at 54% of engine time, and the swiglu fp32-cast semantics that the `npu_dequant_swiglu_quant(quant_mode=1)` fused path replaces should be preserved exactly (the BF16 path intentionally upcasts; a fused int8 path must demonstrate the same numerical intent via the quality gate in point 2).
